### PR TITLE
Add platform info query to identify its backend adapter

### PR DESCRIPTION
**Pull request by smaslov-intel**
This allows users to select wanted adapter/platform.

### CODE REVIEW

**Comment by smaslov-intel** [comment]
@kbenzie : please review/merge

**Comment by smaslov-intel** [comment]
I don't see why there are errors reported in https://github.com/oneapi-src/unified-runtime/actions/runs/4601026787/jobs/8128372658?pr=407

> gmake[3]: *** [CMakeFiles/check-generated.dir/build.make:70: CMakeFiles/check-generated] Error 1
gmake[2]: *** [CMakeFiles/Makefile2:1594: CMakeFiles/check-generated.dir/all] Error 2
gmake[1]: *** [CMakeFiles/Makefile2:1601: CMakeFiles/check-generated.dir/rule] Error 2
gmake: *** [Makefile:595: check-generated] Error 2
Error: Process completed with exit code 2.

**Comment by igchor** [comment]
> I don't see why there are errors reported in https://github.com/oneapi-src/unified-runtime/actions/runs/4601026787/jobs/8128372658?pr=407
> 
> > gmake[3]: *** [CMakeFiles/check-generated.dir/build.make:70: CMakeFiles/check-generated] Error 1
> > gmake[2]: *** [CMakeFiles/Makefile2:1594: CMakeFiles/check-generated.dir/all] Error 2
> > gmake[1]: *** [CMakeFiles/Makefile2:1601: CMakeFiles/check-generated.dir/rule] Error 2
> > gmake: *** [Makefile:595: check-generated] Error 2
> > Error: Process completed with exit code 2.

@smaslov-intel you need to include generated sources in the PR: you should run `make generate` locally and commit all changes.

**Comment by smaslov-intel** [comment]
> > I don't see why there are errors reported in https://github.com/oneapi-src/unified-runtime/actions/runs/4601026787/jobs/8128372658?pr=407
> > > gmake[3]: *** [CMakeFiles/check-generated.dir/build.make:70: CMakeFiles/check-generated] Error 1
> > > gmake[2]: *** [CMakeFiles/Makefile2:1594: CMakeFiles/check-generated.dir/all] Error 2
> > > gmake[1]: *** [CMakeFiles/Makefile2:1601: CMakeFiles/check-generated.dir/rule] Error 2
> > > gmake: *** [Makefile:595: check-generated] Error 2
> > > Error: Process completed with exit code 2.
> 
> @smaslov-intel you need to include generated sources in the PR: you should run `make generate` locally and commit all changes.

Thanks @igchor !
But `make generate` produced tons of changes:
> $ git commit -asm "make generate"
[main 8b23094] make generate
 12 files changed, 31460 insertions(+), 29538 deletions(-)

Is the correct flow described anywhere?

File: include/ur_api.h [hunk 1]
```diff
@@ -544,10 +555,24 @@ urPlatformCreateWithNativeHandle(
...
+    );
+
+///////////////////////////////////////////////////////////////////////////////
+/// @brief Identifies native backend adapters
+typedef enum ur_platform_backend_t
+{
+    UR_PLATFORM_BACKEND_UNKNOWN = 0,                ///< The backend is not a recognized one
+    UR_PLATFORM_BACKEND_LEVEL_ZERO = 1,             ///< The backend is Level Zero
+    UR_PLATFORM_BACKEND_CUDA = 2,                   ///< The backend is CUDA
+    UR_PLATFORM_BACKEND_HIP = 3,                    ///< The backend is HIP
```
Comment (jandres742) [inline include/ur_api.h:570]: OpenCL could be added now as well.

Comment (smaslov-intel) [reply include/ur_api.h]: Sure, added.

**Comment by smaslov-intel** [comment]
> > > I don't see why there are errors reported in https://github.com/oneapi-src/unified-runtime/actions/runs/4601026787/jobs/8128372658?pr=407
> > > > gmake[3]: *** [CMakeFiles/check-generated.dir/build.make:70: CMakeFiles/check-generated] Error 1
> > > > gmake[2]: *** [CMakeFiles/Makefile2:1594: CMakeFiles/check-generated.dir/all] Error 2
> > > > gmake[1]: *** [CMakeFiles/Makefile2:1601: CMakeFiles/check-generated.dir/rule] Error 2
> > > > gmake: *** [Makefile:595: check-generated] Error 2
> > > > Error: Process completed with exit code 2.
> > 
> > 
> > @smaslov-intel you need to include generated sources in the PR: you should run `make generate` locally and commit all changes.
> 
> Thanks @igchor ! But `make generate` produced tons of changes:
> 
> > $ git commit -asm "make generate"
> > [main [8b23094](https://github.com/oneapi-src/unified-runtime/commit/8b230941f848c410f10dea2e36737d5415bac0ce)] make generate
> > 12 files changed, 31460 insertions(+), 29538 deletions(-)
> 
> Is the correct flow described anywhere?

It looks a different clang-format

**Comment by igchor** [comment]
> > > I don't see why there are errors reported in https://github.com/oneapi-src/unified-runtime/actions/runs/4601026787/jobs/8128372658?pr=407
> > > > gmake[3]: *** [CMakeFiles/check-generated.dir/build.make:70: CMakeFiles/check-generated] Error 1
> > > > gmake[2]: *** [CMakeFiles/Makefile2:1594: CMakeFiles/check-generated.dir/all] Error 2
> > > > gmake[1]: *** [CMakeFiles/Makefile2:1601: CMakeFiles/check-generated.dir/rule] Error 2
> > > > gmake: *** [Makefile:595: check-generated] Error 2
> > > > Error: Process completed with exit code 2.
> > 
> > 
> > @smaslov-intel you need to include generated sources in the PR: you should run `make generate` locally and commit all changes.
> 
> Thanks @igchor ! But `make generate` produced tons of changes:
> 
> > $ git commit -asm "make generate"
> > [main [8b23094](https://github.com/oneapi-src/unified-runtime/commit/8b230941f848c410f10dea2e36737d5415bac0ce)] make generate
> > 12 files changed, 31460 insertions(+), 29538 deletions(-)
> 
> Is the correct flow described anywhere?

I think there is only this line in the README: `If you've made modifications to the specification, you can also run a custom generate target prior to building.`

**Comment by igchor** [comment]
> > > > I don't see why there are errors reported in https://github.com/oneapi-src/unified-runtime/actions/runs/4601026787/jobs/8128372658?pr=407
> > > > > gmake[3]: *** [CMakeFiles/check-generated.dir/build.make:70: CMakeFiles/check-generated] Error 1
> > > > > gmake[2]: *** [CMakeFiles/Makefile2:1594: CMakeFiles/check-generated.dir/all] Error 2
> > > > > gmake[1]: *** [CMakeFiles/Makefile2:1601: CMakeFiles/check-generated.dir/rule] Error 2
> > > > > gmake: *** [Makefile:595: check-generated] Error 2
> > > > > Error: Process completed with exit code 2.
> > > 
> > > 
> > > @smaslov-intel you need to include generated sources in the PR: you should run `make generate` locally and commit all changes.
> > 
> > 
> > Thanks @igchor ! But `make generate` produced tons of changes:
> > > $ git commit -asm "make generate"
> > > [main [8b23094](https://github.com/oneapi-src/unified-runtime/commit/8b230941f848c410f10dea2e36737d5415bac0ce)] make generate
> > > 12 files changed, 31460 insertions(+), 29538 deletions(-)
> > 
> > 
> > Is the correct flow described anywhere?
> 
> It looks a different clang-format

Yes, it looks like a wrong clang-format version, but it's strange - we check the specific version in CMake and provide our own rules. What version do you have installed?

I've pushed regenerated sources formatted with clang-15, the diff is much smaller.

**Comment by smaslov-intel** [comment]
> > > > > I don't see why there are errors reported in https://github.com/oneapi-src/unified-runtime/actions/runs/4601026787/jobs/8128372658?pr=407
> > > > > > gmake[3]: *** [CMakeFiles/check-generated.dir/build.make:70: CMakeFiles/check-generated] Error 1
> > > > > > gmake[2]: *** [CMakeFiles/Makefile2:1594: CMakeFiles/check-generated.dir/all] Error 2
> > > > > > gmake[1]: *** [CMakeFiles/Makefile2:1601: CMakeFiles/check-generated.dir/rule] Error 2
> > > > > > gmake: *** [Makefile:595: check-generated] Error 2
> > > > > > Error: Process completed with exit code 2.
> > > > 
> > > > 
> > > > @smaslov-intel you need to include generated sources in the PR: you should run `make generate` locally and commit all changes.
> > > 
> > > 
> > > Thanks @igchor ! But `make generate` produced tons of changes:
> > > > $ git commit -asm "make generate"
> > > > [main [8b23094](https://github.com/oneapi-src/unified-runtime/commit/8b230941f848c410f10dea2e36737d5415bac0ce)] make generate
> > > > 12 files changed, 31460 insertions(+), 29538 deletions(-)
> > > 
> > > 
> > > Is the correct flow described anywhere?
> > 
> > 
> > It looks a different clang-format
> 
> Yes, it looks like a wrong clang-format version, but it's strange - we check for specific version in CMake and provide our own rules. What version do you have installed (if any)?
> 
> I've pushed regenerated sources formatted with clang-15, the diff is much smaller.

$ clang-format --version
clang-format version 17.0.0 (git@github.com:smaslov-intel/llvm.git 7719738e32500bff5a61193bfcff9c978ca52e88)

**Comment by igchor** [comment]
Ok, I think we didn't actually check for clang-version when doing 'generate': https://github.com/oneapi-src/unified-runtime/pull/408

**Comment by smaslov-intel** [comment]
> Is this information not already available via the `UR_PLATFORM_INFO_NAME`?
> 
> I'm wary about the extensibility of an enum.

The `UR_PLATFORM_INFO_NAME` returns a string, so we'd need a spec how platforms are named so that runtimes can map it back to what they can work with (enums). Rather than having that extra spec and map in each user, I think it is preferable to just add enum here.

**Comment by smaslov-intel** [comment]
> > The `UR_PLATFORM_INFO_NAME` returns a string, so we'd need a spec how platforms are named so that runtimes can map it back to what they can work with (enums). Rather than having that extra spec and map in each user, I think it is preferable to just add enum here.
> 
> Okay, while duplicating information isn't ideal I can accept it. Where do you expect this to be used, is this passed through the SYCL RT up to the application?

The immediate use is for SYCL users to select the underlying backend (via ONEAPI_DEVICE_SELECTOR). Today this selection is via choosing a PI plugin hardcoded in SYCL RT, and with our move to UR we need this new functionality since no longer have visibility into UR adapters.

**Comment by kbenzie** [comment]
> The immediate use is for SYCL users to select the underlying backend (via ONEAPI_DEVICE_SELECTOR). Today this selection is via choosing a PI plugin hardcoded in SYCL RT, and with our move to UR we need this new functionality since no longer have visibility into UR adapters.

Ah I see, so its related to https://github.com/oneapi-src/unified-runtime/issues/220. When we discussed this in the WG we came to the conclusion this logic would get pulled into UR once PI has been removed. Perhaps when that happens this query would no longer be necessary.

**Review by kbenzie — APPROVED** [review]

**Comment by smaslov-intel** [comment]
> > The immediate use is for SYCL users to select the underlying backend (via ONEAPI_DEVICE_SELECTOR). Today this selection is via choosing a PI plugin hardcoded in SYCL RT, and with our move to UR we need this new functionality since no longer have visibility into UR adapters.
> 
> Ah I see, so its related to #220. When we discussed this in the WG we came to the conclusion this logic would get pulled into UR once PI has been removed. Perhaps when that happens this query would no longer be necessary.

Unless all the device filtering is moved into UR we will need this interface to stay. Note, SYCL language has intrinsic notion of "backend", so I think this will have to stay forever. E.g. see https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_oneapi_filter_selector.asciidoc

**Comment by smaslov-intel** [comment]
@kbenzie : can this be merged, please?

**Review by veselypeta — APPROVED** [review]
LGTM

**Comment by kbenzie** [comment]
> @kbenzie : can this be merged, please?

Sure. We aim for two approvals so I was waiting for another reviewer.

**Pull request merged**